src/vmm/src/devices/virtio/queue.rs

```diff
@@ -562,25 +562,87 @@ impl Queue {
     };
-    use crate::vstate::memory::{
-        Address, AtomicBitmap, Bytes, FileOffset, GuestAddress, GuestMemory, GuestMemoryMmap,
-        GuestRegionMmap, MmapRegion,
-    };
+    use crate::vstate::memory::{Bytes, FileOffset, GuestAddress, GuestMemory, MmapRegion};
+
+    /// A made-for-kani version of `vm_memory::GuestMemoryMmap`. Unlike the real
+    /// `GuestMemoryMmap`, which manages a list of regions and then does a binary
+    /// search to determine which region a specific read or write request goes to,
+    /// this only uses a single region. Eliminating this binary search significantly
+    /// speeds up all queue proofs, because it eliminates the only loop contained herein,
+    /// meaning we can use `kani::unwind(0)` instead of `kani::unwind(2)`. Functionally,
+    /// it works identically to `GuestMemoryMmap` with only a single contained region.
+    pub struct ProofGuestMemory {
+        the_region: vm_memory::GuestRegionMmap,
+    }
+
+    impl<'a> GuestMemoryIterator<'a, vm_memory::GuestRegionMmap> for ProofGuestMemory {
+        type Iter = std::iter::Once<&'a vm_memory::GuestRegionMmap>;
+    }
 
-    pub struct ProofContext(pub Queue, pub GuestMemoryMmap);
+    impl GuestMemory for ProofGuestMemory {
+        type R = vm_memory::GuestRegionMmap;
+        type I = Self;
+
+        fn num_regions(&self) -> usize {
+            1
+        }
+
+        fn find_region(&self, addr: GuestAddress) -> Option<&Self::R> {
+            self.the_region
+                .to_region_addr(addr)
+                .map(|_| &self.the_region)
+        }
+
+        fn iter(&self) -> <Self::I as GuestMemoryIterator<Self::R>>::Iter {
+            std::iter::once(&self.the_region)
+        }
+
+        fn try_access<F>(
+            &self,
+            count: usize,
+            addr: GuestAddress,
+            mut f: F,
+        ) -> vm_memory::guest_memory::Result<usize>
+        where
+            F: FnMut(
+                usize,
+                usize,
+                MemoryRegionAddress,
+                &Self::R,
+            ) -> vm_memory::guest_memory::Result<usize>,
+        {
+            // We only have a single region, meaning a lot of the complications of the default
+            // try_access implementation for dealing with reads/writes across multiple
+            // regions does not apply.
+            let region_addr = self
+                .the_region
+                .to_region_addr(addr)
+                .ok_or(vm_memory::guest_memory::Error::InvalidGuestAddress(addr))?;
+            self.the_region
+                .checked_offset(region_addr, count)
+                .ok_or(vm_memory::guest_memory::Error::InvalidGuestAddress(addr))?;
+            f(0, count, region_addr, &self.the_region)
+        }
+    }
+
+    pub struct ProofContext(pub Queue, pub ProofGuestMemory);
 
     pub struct MmapRegionStub {
         addr: *mut u8,
         size: usize,
-        bitmap: Option<AtomicBitmap>,
+        bitmap: (),
```

Comment:
Mh, let me check. If it doesnt affect performance then I don't see why we shouldn't add it back tbh. I think I removed it in prototyping and then just forget to add it back

Reply:
If it doesn't add overhead, I would include it anyway. Even if we don't exercise related functionality today, we might do in the future.

Reply:
Mh, seems it does add quite some overhead. I had some look at the code in question, and it contains some non-trivial arithmetic and loop that causes kani to go back to not terminating. Which now leaves me to wonder if maybe this is the real reason why performance improved instead of mocking out `GuestMemoryMmap` :thinking:

Reply:
Nope, both optimizations are needed

Reply:
what about the question whether we need this?

Reply:
I think its orthogonal to the properties we want to prove here: the bitmap tracking is vm-memory functionality and should be covered by kani proofs in vm-memory. Here we want to prove correctness properties of the virtio queue implementation, which are independent on the memory tracking functionality.